### ISSUE TITLE
Free-threaded support?

### ISSUE THREAD

**Issue by clin1234** [mhammond/pywin32#2303]
Has anyone here given thought on testing this module with the no-GIL version of the Python interpreter yet?

**Comment by mhammond** [comment]
I'm not aware of anyone looking at this. It's worth noting that there are quite a few places where the GIL is used to prevent concurrent access, so identifying these seems like the first step.

**Comment by Avasam** [comment]
As far as CI builds go,  would be nice. But if it's somehow not available by the time pywin32 has nogil support, deadsnakes already offer a nogil python action: https://github.com/deadsnakes/action

**Comment by clin1234** [comment]
> As far as CI builds go, [actions/setup-python#771](https://github.com/actions/setup-python/issues/771) would be nice. But if it's somehow not available by the time pywin32 has nogil support, deadsnakes already offer a nogil python action: https://github.com/deadsnakes/action

deadsnakes is only tested on Ubuntu

**Comment by Avasam** [comment]
> As far as CI builds go, [actions/setup-python#771](https://github.com/actions/setup-python/issues/771) would be nice.

It's also possible uv gets to it first: https://github.com/astral-sh/uv/issues/7193

**Comment by Avasam** [comment]
Here's a workflow manually installing Python 3.13t: https://github.com/mhammond/pywin32/pull/2395
It builds, but as expected, tests don't get very far:
> --- Running 'D:\a\pywin32\pywin32\win32/test/testall.py' ---
> *** Test script 'D:\a\pywin32\pywin32\win32/test/testall.py' exited with 3221225477
> --- Running 'D:\a\pywin32\pywin32\Pythonwin/pywin/test/all.py' ---
> Fatal Python error: PyWinInterpreterState_Ensure: pywintypes: can not setup interpreter state, as current state is invalid
> Python runtime state: unknown

> 
> *** Test script 'D:\a\pywin32\pywin32\Pythonwin/pywin/test/all.py' exited with 3221226505
> --- Running 'D:\a\pywin32\pywin32\com\win32com\test\testall.py' ---
> *** Test script 'D:\a\pywin32\pywin32\com\win32com\test\testall.py' exited with 3221225477
> The following scripts failed
> \> D:\a\pywin32\pywin32\win32/test/testall.py
> \> D:\a\pywin32\pywin32\Pythonwin/pywin/test/all.py
> \> D:\a\pywin32\pywin32\com\win32com\test\testall.py

This is unfortunately not something I'll be able to provide much help with. Contributions are more than welcome !

**Comment by vivek-gite** [comment]
@mhammond Do we have any plan to release free threaded version of pywin32 ?

**Comment by mhammond** [comment]
> Hi [@mhammond](https://github.com/mhammond) Do we have any plan to release free threaded version of pywin32 ?

No, I've no plans to work on this and I've heard no one else say they will, and even if they do, I suspect the patches will be too large and risky to accept.

**Comment by 0xDEADFED5** [comment]
=(

> > Hi [@mhammond](https://github.com/mhammond) Do we have any plan to release free threaded version of pywin32 ?
> 
> No, I've no plans to work on this and I've heard no one else say they will, and even if they do, I suspect the patches will be too large and risky to accept.

**Comment by Avasam** [comment]
@mhawash You are more than welcome to provide patches as Pull Requests to work towards Free-Threaded support for pywin32. Or find someone who would. There is a lot of unknowns on the way there, and we expect it to be an arduous road. I don't know if partial support would be accepted either (see similar discussions for PyPy https://github.com/mhammond/pywin32/issues/1289). I'm somewhat open to it, as I have projects that use pywin32 w/o using COM objects for example. 

https://github.com/mhammond/pywin32/issues/2303#issuecomment-2186790621
> It's worth noting that there are quite a few places where the GIL is used to prevent concurrent access, so identifying these seems like the first step.

https://github.com/mhammond/pywin32/issues/2303#issuecomment-2607368209
> No, I've no plans to work on this and I've heard no one else say they will, and even if they do, I suspect the patches will be too large and risky to accept.

https://github.com/jupyterlab/jupyterlab/issues/17887#issuecomment-3341995611
> > Also, for what it’s worth I don’t think the pywin32 maintainer said they won’t support free-threaded python, just that no one has worked on it yet and they suspect it will be a big task.
> 
> Correct. It's not that we don't want to support FreeThreaded Python (I myself have an app that relies on pywin32 that I would love to see use FreeThreaded Python eventually). It's moreso that none of the current (3) pywin32 maintainers have the will, free-threaded expertise, and/or free time to look into and learn everything necessary to make it work. Especially given that pywin32 actively relies on the GIL in a handful of places, so we're expecting a lot of pain and risks in a potential migration.
> 
> So if it happens, it's likely to be from external contributors.

**Comment by Avasam** [comment]
(Keeping this open since just getting the CI to build is still some ways away from actual support, as mentionned in https://github.com/mhammond/pywin32/pull/2767#pullrequestreview-4445160405 and https://github.com/mhammond/pywin32/blob/main/NOGIL.md )

Thanks @clin1234 !

Feel free to build from source, test, and provide patches 🙂
I also spent a lot of effort cleaning up the build process and old cruft to make that easier and faster.
